virtual_rainforest/soil/model.py

```diff
@@ -0,0 +1,67 @@
+"""The `soil.model` module.
+
+The `soil.model` module creates a `SoilModel` class, which extended the base `Model`
+class to be usable to simulate the soil.
+"""
+
+from numpy import datetime64, timedelta64
+
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+from virtual_rainforest.core.model import BaseModel
+
+
+class InitialisationError(Exception):
+    """Custom exception class for model initialisation failures."""
+
+    pass
+
+
+class SoilModel(BaseModel):
+    """A class describing the soil model.
+
+    Describes the specific functions and attributes that the soil module should possess.
+    It's very much incomplete at the moment, and only overwrites one function in a
+    pretty basic manner. This is intended as a demonstration of how the class
+    inheritance should be handled for the model classes.
+
+    Args:
+        start_time: Point in time that the model simulation should be started.
+        end_time: Time that the model simulation should end
+        update_interval: Time to wait between updates of the model state.
+
+    Attributes:
+        name: Names the model that is described
+    """
+
+    name = "soil model"
+
+    # THIS IS BASICALLY JUST A PLACEHOLDER TO DEMONSTRATE HOW THE FUNCTION OVERWRITING
+    # SHOULD WORK
+    def setup(self) -> None:
+        """Function to set up the soil model."""
+        for layer in range(0, self.no_layers):
+            LOGGER.info(f"Setting up soil layer {layer}")
+
+    def __repr__(self) -> str:
+        """Represent a Model as a string."""
+        return (
+            f"SoilModel(start_time={self.start_time}, end_time={self.end_time}, "
+            f"update_interval={self.update_interval}, no_layers={self.no_layers})"
+        )
+
+    def __init__(
+        self,
+        start_time: datetime64,
+        end_time: datetime64,
+        update_interval: timedelta64,
+        no_layers: int,
```

Comment:
Given that this child class is changing the signature of the parent class - `BaseClass` has no `no_layers` -, I'd use in `*args, **kwargs` in the `__init__` method of the later so it is clear that child classes might have more positional and keyword arguments. I'd keep those two here, in case this class is, in time, subclassed with more specialised versions.

Reply:
So `SoilModel.__init__` would take `*args, **kwargs` which would then be unpacked and used to init `BaseModel`? Or should every `.__init__` method take `*args, **kwargs` as arguments?

Reply:
At least the base model - even if they are not used there - so it becomes clear that the signature of child classes might have more arguments. Changing the signature of methods in child classes is generally considered bad practice. Constructors are slightly different because it is common to have child classes requiring different inputs than the base classes, but even in that case, it is good if it can be indicated that the signature could be fluid in children. See for example https://stackoverflow.com/a/64156492

Reply:
At some point in the program structure, we need to be unpacking chunks of the `config` into their appropriate models. Now, we could do that:

1.  by having those config options as arguments to the subclasses (and hence use `*args, **kwargs` in the base class and call a subclass using `Plants(**config['plants'])`) or,
2.  we could have the base class accept a `config` arguments so that there is always a single argument to accept a config object that then gets unpacked within `__init__` (so models are always created as `Plants(config=config['plants'])`.

Having typed that out - I don't much like the second option. The first is clearer because then we have to document the subclass `__init__` arguments in the class docstring, which is the natural place for them. The only advantage I can see is that some of these models could have very complex signatures?